Add unique index to object (bucket_id, name)

irq0 opened the pull request:
A bucket cannot have the same object key twice. Use unique index to
enforce this. Adapt create version path to use existing objects having deleted versions only.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

Commit:
rgw/sfs: Add unique index to object (bucket_id, name)

A bucket cannot have the same object key twice. Use unique index to
enforce this.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

Commit:
rgw/sfs: create_new_versioned: Find uuid even with deleted versions

Remove deleted version filter. Even with deleted versions, we must not
create a new object entry as the (name, bucket_id) combination must be
unique. As this is enforced by a previous change, the function no
longer succeeds if deleted versions exists. It will loop around trying
to create a new object row until retries are used up.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>

jecluis reviewed (dismissed):
This would benefit from having an upgrade path story, but if it happens to be too convoluted we need to make sure we do tell users that their store is no longer compatible.

irq0 commented:
Last push pulls the unique index spec "freestanding". This will make .sync actually call it. Having it in the table spec will not.

This will cause the .sync fail if a incompat database is found:

```
2023-08-11T11:39:35.163+0200 7f537b6d8680 -1 [SQLITE PROFILE] 0x557462664380 0ms "PRAGMA user_version"                                                                     2023-08-11T11:39:35.163+0200 7f537b6d8680 10 db user version: 3
2023-08-11T11:39:35.166+0200 7f537b6d8680 -1 [SQLITE] (11) database corruption at line 74186 of [831d0fb283]
2023-08-11T11:39:35.166+0200 7f537b6d8680 -1 [SQLITE] (11) statement aborts at 14: [CREATE UNIQUE INDEX IF NOT EXISTS 'object_bucketid_name' ON 'objects' ("bucket_id", "na
me")] database disk image is malformed
2023-08-11T11:39:35.166+0200 7f537b6d8680 -1 [SQLITE] (28) file unlinked while open: /compile/tmp/sfs/s3gw.db_tmp
2023-08-11T11:39:35.166+0200 7f537b6d8680  0 rgw main: ERROR ACCESSING SFS METADATA. Metadata database might be corrupted or is no longer compatible
2023-08-11T11:39:35.166+0200 7f537b6d8680 -1 Couldn't init storage provider (RADOS)
```

Since we want https://github.com/aquarist-labs/s3gw/issues/668 in the release, I'd rather not do conversions and settle for the error message.

jecluis commented:
I think we should then increase the db version and min supported version.

Commit:
rgw/sfs: db version 4 - min compat version 4

As the previous changes a potentially incompatible if duplicates
exists, bump db version to 4. Since we planned data store changes that
won't get an upgrade path require min compat 4 as well.

Signed-off-by: Marcel Lauhoff <marcel.lauhoff@suse.com>
